docs/docs/tutorials/cache/index.md

```diff
@@ -0,0 +1,185 @@
+        {write your logic of computing cache key}
+
+    def get(self, request: Dict[str, Any], ignored_args_for_cache_key: Optional[list[str]] = None) -> Any:
+        {write your cache read logic}
+
+    def put(
+        self,
+        request: Dict[str, Any],
+        value: Any,
+        ignored_args_for_cache_key: Optional[list[str]] = None,
+        enable_memory_cache: bool = True,
+    ) -> None:
+        {write your cache write logic}
+```
+
+To ensure seamless integration with the rest of DSPy, it is recommended to implement your custom cache using the same method signatures as the base class, or at a minimum, include `**kwargs` in your method definitions to prevent runtime errors during cache read/write operations.
+
+Once your custom cache class is defined, you can instruct DSPy to use it:
+
+```python
+dspy.cache = CustomCache()
+```
+
+Let's illustrate this with a practical example. Suppose we want the cache key computation to depend solely on the request message content, ignoring other parameters like the specific LM being called. We can create a custom cache as follows:
+
+```python
+class CustomCache(dspy.clients.Cache):
+
+    def cache_key(self, request: Dict[str, Any], ignored_args_for_cache_key: Optional[list[str]] = None) -> str:
+        messages = request.get("messages", [])
+        return sha256(ujson.dumps(messages, sort_keys=True).encode()).hexdigest()
+
+dspy.cache = CustomCache(enable_disk_cache=True, enable_memory_cache=True, disk_cache_dir=dspy.clients.DISK_CACHE_DIR)
+```
+
+For comparison, consider executing the code below without the custom cache:
+
+```python
+import dspy
+import os
+import time
+
+os.environ["OPENAI_API_KEY"] = "{your_openai_key}"
+
+dspy.settings.configure(lm=dspy.LM("openai/gpt-4o-mini"))
+
+predict = dspy.Predict("question->answer")
+
+start = time.time()
+result1 = predict(question="Who is the GOAT of soccer?")
+print(f"Time elapse: {time.time() - start: 2f}")
+
+start = time.time()
+with dspy.context(lm=dspy.LM("openai/gpt-4.1-mini")):
+    result2 = predict(question="Who is the GOAT of soccer?")
+print(f"Time elapse: {time.time() - start: 2f}")
+```
+
+The time elapsed will indicate that the cache is not hit on the second call. However, when using the custom cache:
+
+```python
+import dspy
+import os
+import time
+from typing import Dict, Any, Optional
+import ujson
+from hashlib import sha256
+
+os.environ["OPENAI_API_KEY"] = "{your_openai_key}"
+
+dspy.settings.configure(lm=dspy.LM("openai/gpt-4o-mini"))
+
+class CustomCache(dspy.clients.Cache):
+
+    def cache_key(self, request: Dict[str, Any], ignored_args_for_cache_key: Optional[list[str]] = None) -> str:
+        messages = request.get("messages", [])
+        return sha256(ujson.dumps(messages, sort_keys=True).encode()).hexdigest()
+
+dspy.cache = CustomCache(enable_disk_cache=True, enable_memory_cache=True, disk_cache_dir=dspy.clients.DISK_CACHE_DIR)
```

Comment:
q: should we make this configurable through `dspy.settings.configure` for consistency?

Reply:
`cache` is more like a global singleton, while `dspy.settings` are supposed to be configured for individual working threads. and btw, I feel that `dspy.settings` has been abused as a data carrier and has become too big, we may want to delete some fields?

Reply:
Think about dspy.Evaluate - when we launch evaluation we are spinning up multiple working threads, but we still want to reuse the cache to save budget and speed up future queries. In most cases cache is something users shouldn't need to be aware of, while `dspy.settings` are meant to be configurable by users.